include/deal.II/fe/fe_enriched.h

```diff
@@ -0,0 +1,545 @@
+#include <deal.II/fe/fe.h>
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/fe_nothing.h>
+#include <deal.II/fe/fe_update_flags.h>
+#include <deal.II/base/function.h>
+#include <deal.II/base/symmetric_tensor.h>
+#include <deal.II/base/quadrature.h>
+
+#include <vector>
+#include <utility>
+#include <numeric>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * Implementation of a partition of unity finite element which enriches a standard
+ * finite element with an enrichment function multiplied with another (usually
+ * linear) finite element:
+ * \f[
+ * U(x) = \sum_i N_i(x) U_i + \sum_j N_j(x) \sum_k F_k(x) U_{jk}
+ * \f]
+ * where \f$ N_i(x) \f$ and \f$ N_j(x) \f$ are the underlying finite elements (including
+ * the mapping from the isoparametric element to the real element); \f$ F_k(x) \f$
+ * are the enrichment functions in real space (e.g. \f$ 1/r \f$, \f$ exp(-r) \f$, Heaviside, etc);
+ * \f$ U_i \f$ and \f$ U_{jk} \f$ are the standard and enriched DoFs.
+ *
+ * The simplest constructor of this class takes two finite element objects and an
+ * enrichment function to be used. For example
+ *
+ * @code
+ * FE_Enriched<dim> fe(FE_Q<dim>(1),
+ *                     FE_Q<dim>(1),
+ *                     function)
+ * @endcode
+ *
+ * The \f$ exp(-x) \f$ enrichment leads to the following shape functions on unit
+ * element:
+ * <table>
+ *   <tr>
+ *     <td align="center">
+ *       @image html fe_enriched_1d.png
+ *     </td>
+ *     <td align="center">
+ *       @image html fe_enriched_h-refinement.png
+ *     </td>
+ *   </tr>
+ *   <tr>
+ *     <td align="center">
+ *       1d element, base and partition-of-unity shape functions.
+ *     </td>
+ *     <td align="center">
+ *       enriched shape function resulting from h-refinement of a single element (bottom-left corner).
+ *     </td>
+ *   </tr>
+ * </table>
+ *
+ * <h3>Implementation</h3>
+ *
+ * The implementation of the class is based on FESystem which is aggregated as
+ * a private member and initialized for the case above as
+ *
+ * @code
+ * FESystem<dim> fe_system(FE_Q<dim>(1),1,
+ *                         FE_Q<dim>(1),1);
+ * @endcode
+ *
+ * Note that it would not be wise to have this class derived
+ * from FESystem as the latter concatenates the given elements into different
+ * components of a vector element, whereas the current class combines the given
+ * elements into the same components. For instance, if two scalar elements are
+ * given, the resulting element will be scalar rather than have two components
+ * when doing the same with an FESystem.
+ *
+ * The ordering of the shape function, @p interface_constrains, the @p prolongation (embedding)
+ * and the @p restriction matrices are taken from the FESystem class.
+ *
+ * <h3>hp-constraints</h3>
+ *
+ * This class implements hp constraints with arbitrary FE. For the example above,
+ * a user will need to wrap a given FiniteElement into FE_Enriched:
+ * @code
+ *   FE_Enriched<dim> fe_enriched(FE_Q<dim>(1));
```

Comment:
> Oh, so are you saying that one would use the wrapper because this allows to use FE_Enriched(FE_Q) on one cell, and FE_Enriched(FE_Q, FE_Q, function) on another and automatically get constraints working, whereas the same would not be true for FE_Q and FE_Enriched(FE_Q, FE_Q, function)?

exactly! The point is to reuse `FESystem` as much as possible.

> If this is so, then I think this is unnecessarily complicated. Just force users to use FE_Enriched (FE_Q, FE_Nothing, something) instead. This saves you a bunch of code and makes things a lot simpler. 

It could make more difficult for users as they need to be aware to use `FE_Nothing(dominate)` to get constraints working. But I can of course add checks to constructors that this is the case.

On another hand, i agree with your that having two constructors `FE_Enriched(FE_Q, FE_Q, function)` and the complicated general one below is more logical.